### ISSUE TITLE
Convert Java Iterator classes to implement IEnumerable<T>

### ISSUE THREAD

**Issue by bongohrtech** [bongohrtech/lucenenet#137]
<p>The Iterator pattern in Java is equivalent to IEnumerable in .NET.  Classes that were directly ported in Java using the Iterator pattern, cannot be used with Linq or foreach blocks in .NET.</p>

<p><tt>Next()</tt> would be equivalent to .NET's <tt>MoveNext()</tt>, and in the below case, <tt>Term()</tt> would be as .NET's <tt>Current</tt> property.  In cases as below, it will require <tt>TermEnum</tt> to become an abstract class with <tt>Term</tt> and <tt>DocFreq</tt> properties, which would be returned from another class or method that implemented <tt>IEnumerable<TermEnum></tt>.</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
	public abstract class TermEnum : IDisposable
	{
		public abstract bool Next();
		public abstract Term Term();
		public abstract int DocFreq();
		public abstract void  Close();
	public abstract void Dispose();
	}
</pre>
</div></div> 

<p>would instead look something like:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
	public class TermFreq
	{
		public abstract Term { get; }
		public abstract int { get; }
	}

public abstract class TermEnum : IEnumerable<TermFreq>, IDisposable
{
// ...
}
</pre>
</div></div>

<p>Keep in mind that it is important that if the class being converted implements <tt>IDisposable</tt>, the class that is enumerating the terms (in this case <tt>TermEnum</tt>) should inherit from both <tt>IEnumerable<T></tt> <b>and</b> <tt>IDisposable</tt>.  This won't be any change to the user, as the compiler automatically calls <tt>IDisposable</tt> when used in a <tt>foreach</tt> loop.</p>
<i>JIRA link - <a href="https://issues.apache.org/jira/browse/LUCENENET-469">[LUCENENET-469]</a> created by ccurrens</i>

**Comment by bongohrtech** [comment]
<p>This requires too much of an API change to implement in 3.0.3.  Unassigning from 3.0.3, and we can revisit it in a later release.</p>
<i>by ccurrens</i>

**Comment by bongohrtech** [comment]
<p>We should probably look into doing this. There are several places in the code base where the dispose pattern is not used correctly in conjunction with iterators/enumerators.</p>

<p>I am not sure if it makes sense to make every type of iterator into an enumerator. For example BreakIterator is designed to navigate forward, backward, and to skip to specific points. Perhaps we could still implement IEnumerable<T> but also create other interfaces with extra functionality. In any case, this is a major API change that should be looked at before the 4.8.0 release.</p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<p>How about something like...</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">	<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> <span class="code-keyword">class </span>TermEnum : IEnumerable<TermFreq>, IDisposable
	{
		<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> bool Next();
		<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> Term Term();
		<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> <span class="code-object">int</span> DocFreq();
		<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> void Close();
		<span class="code-keyword">public</span> <span class="code-keyword">abstract</span> void Dispose();

		<span class="code-keyword">public</span> IEnumerator<TermFreq> GetEnumerator() => <span class="code-keyword">new</span> EnumEnumerator<TermFreq>(Next, Dispose, () => <span class="code-keyword">new</span> TermFreq { Term = Term(), DocFreq = DocFreq() });

		IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
	}

	<span class="code-keyword">public</span> <span class="code-keyword">class </span>TermFreq
	{
		<span class="code-keyword">public</span> Term Term { get; set; }
		<span class="code-keyword">public</span> <span class="code-object">int</span> DocFreq { get; set; }
	}

	<span class="code-keyword">public</span> <span class="code-keyword">class </span>EnumEnumerator<T> : IEnumerator<T>
	{
		<span class="code-keyword">private</span> readonly Func<bool> next;
		<span class="code-keyword">private</span> readonly Action dispose;
		<span class="code-keyword">private</span> readonly Func<T> currentFactory;

		<span class="code-keyword">private</span> bool started = <span class="code-keyword">false</span>;

		<span class="code-keyword">public</span> EnumEnumerator(Func<bool> next, Action dispose, Func<T> currentFactory)
		{
			<span class="code-keyword">this</span>.next = next;
			<span class="code-keyword">this</span>.dispose = dispose;
			<span class="code-keyword">this</span>.currentFactory = currentFactory;
		}

		<span class="code-keyword">public</span> T Current => started ? currentFactory() : <span class="code-keyword">default</span>(T);

		object IEnumerator.Current => Current;

		<span class="code-keyword">public</span> bool MoveNext() => (started = next());

		<span class="code-keyword">public</span> void Reset() => <span class="code-keyword">throw</span> <span class="code-keyword">new</span> NotImplementedException();

		<span class="code-keyword">public</span> void Dispose() => dispose();
	}
</pre>
</div></div>

<p>Excuse the c#7 lambda stuff, just used for brevity. Would expand if theory accepted.</p>

<p>The idea is to have a default generic Enumerator which can be used for any Iterator type. So, simply add the IEnumerable i/f and the two GetEnumerator methods returning the generic. It takes Next, Dispose and a factory to create the items.<br/>
The started member is to emulate the enumerator semantic of being null before the first call to MoveNext.<br/>
Enum's aren't resetable so just throw?<br/>
In this example I've kept the item type simple. But it would probably be implemented as an immutable.</p>

<p>Observation: There don't seem to be many Enum types (those that have <tt>bool Next()</tt>) I'm assuming other have already been converted?</p>

<p>Thoughts?</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<p>The types I am aware of are the <tt>TermsEnum</tt> and <tt>DocIdSetIterator</tt> and their subclasses (which include <tt>DocsEnum</tt> and <tt>DocsAndPositionsEnum</tt> classes). There are probably a few others. Anything that is not public facing isn't worth the effort of refactoring, so there isn't much after these that haven't already been converted.</p>

<p>BTW - there is no longer a <tt>TermEnum</tt> class in Lucene 4+ - it has been replaced with <tt>TermsEnum</tt>.</p>

<p>The issue isn't so much with implementing <tt>IEnumerator<T></tt> as the fact that once you use a foreach loop and it calls <tt>GetEnumerator()</tt>, there would need to be a cast to get to the functionality other than <tt>IEnumerator<T></tt>. And inside of a foreach loop you don't have access to the <tt>IEnumerator<T></tt> instance, so there is no way to cast to get the rest of the functionality. </p>

<p>I looked at <tt>TermsEnum</tt> in some depth and I think the best solution there would be to divide it into "basic" and "advanced" APIs. Simply rename <tt>Terms.GetIterator(BytesRef)</tt> to <tt>Terms.GetEnumerator(BytesRef)</tt> and add an overload that accepts no parameter and satisfies the <tt>IEnumerable<T></tt> contract that internally calls <tt>Terms.GetEnumerator(null)</tt>. The overload that accepts a parameter can then return a <tt>TermsEnum</tt> type, which will expose the additional members that are hidden in the .NETified overload.</p>

<p>Since a common use case is to loop forward, this would simplify some code:</p>

<div class="code panel" style="border-style: solid;border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Terms vector = <something>
TermsEnum termsEnum = vector.GetIterator(<span class="code-keyword">null</span>);
BytesRef text;
<span class="code-keyword">while</span> ((text = termsEnum.Next()) != <span class="code-keyword">null</span>)
{
	<span class="code-comment">// use text
</span>}
</pre>
</div></div>

<p>becomes</p>

<div class="code panel" style="border-style: solid;border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Terms vector = <something>
foreach (BytesRef text in vector)
{
	<span class="code-comment">// use text
</span>}
</pre>
</div></div>

<p>But you could also call <tt>GetEnumerable(BytesRef)</tt> in order to reuse a <tt>BytesRef</tt> or to get to all of <tt>TermsEnum</tt> s goodies that aren't part of the <tt>IEnumerator<T></tt> contract.</p>

<p>I don't think this same approach will work for <tt>DocsEnum</tt> or <tt>DocsAndPositionsEnum</tt> - it would be nice to find a solution that can be applied consistently. I am not so sure those types fit the mold of <tt>IEnumerator<T></tt> anyway, but perhaps there is some way to refactor them to fit.</p>

<p>Then again, there must be some reason why the Lucene designers strayed away from using <tt>Iterable<T></tt> in this part of the design - they may have intentionally done it this way for some odd reason.</p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<p>Oh, one more thing. I was planning on getting to this after the next beta release. There are several bug fixes, optimizations, and 5 new NuGet packages I would like to push out (and hopefully the API docs) before getting to these breaking API changes.</p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<p>So I have a little time to get into this then <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Probably nothing new here to you. Mainly just me catching up...</p>

<p>To your first point about GetEnumerator being somewhat hidden/opaque, isn't that just something that is the usual guidance in c#. ie don't mess with the collection from inside a for cos you'll break the enumerator.</p>

<p>In the fairly simple case in the code above you could...<br/>
(NB just using TermEnum as it's a simple example)</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">TermEnum vector = <something>
foreach (<span class="code-keyword">var</span> item in vector)
{
	<span class="code-keyword">var</span> x = vector.Term();
	<span class="code-keyword">var</span> y = item.Term;  <span class="code-comment">// same thing as <span class="code-quote">'x'</span>
</span>	vector.DoStuff();
	<span class="code-comment">// but you should probably steer clear of using <span class="code-quote">'Next()'</span>
</span>}
</pre>
</div></div>

<p>There's no need to extend/cast or get access to the enumerator. The enumerators job is to simply move a cursor forwards and support a language construct.<br/>
Of course this would make it still look weird to a c#/vb dev.<br/>
<tt>Current</tt> is supposed to return the current value of the cursor. Maybe it could return Current as the collection ('vector' above). So it would look like...</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">TermEnum vector = <something>
foreach (TermEnum item in vector)
{
	<span class="code-keyword">var</span> x = item.Term();
}
</pre>
</div></div>

<p>Again, weird. Doing the method -> property change would help. But, still feels odd to have the item being the same thing as vector.</p>

<p>Here lies the basic problem. The Lucene "iterable" classes represent all of: </p>
<ul>
	<li>the collection
	<ul>
		<li>stepping the cursor</li>
		<li>collection scope behaviors for the type</li>
	</ul>
	</li>
	<li>the current item within the collection
	<ul>
		<li>"properties"</li>
		<li>item scope behaviors<br/>
Not only that but there are also different implementations of the pattern with different semantics. ie <tt>Next()</tt> returns success; <tt>NextDoc()</tt> returns an id or magic number; <tt>Advance(target)</tt> spin forward to target</li>
	</ul>
	</li>
</ul>


<p>Also, the need to allow the Current item to be; created on each iteration; reused; fetch from object pool; ...<br/>
I'm fairly sure that the EnumEnumerator (or versions of) can handle all that. I'd need to get further into the use cases.</p>

<p>Lastly, there's the observation that the <tt>while</tt> vs <tt>foreach</tt> examples you gave are not equivalent as the while version does not call <tt>Dispose()</tt>. <br/>
I suspect that in current usage Dispose is never called. Or if it is it's because it's associated with files rather than because we're leaving a loop. Does that match your experience?<br/>
Thinking about it, maybe the enumerator Dispose should be a noop? It's odd for a type to be disposing it's "parent" (though I know some StreamWriter/Reader classes do, it's still odd/annoying). Again there's a scoping concept here that's muddled by the type being many things.</p>

<p>All very Tricky...</p>

<p>I think there's some combination of</p>
<ul>
	<li>isolating the "item" type</li>
	<li>using the collection object behaviors from inside the loop</li>
	<li>implementing the enumerator such that it is ok about the cursor being changed from outside</li>
</ul>


<p>It's going to be very hard to get the semantics of each type in a sane shape such that it's hard for a dev to get tripped up.<br/>
Which is why this have been open for 5+ years <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I'll see if I can put together a proof of concept PR</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<p>No problem. I am just a bit surprised given the choice for some prime green field development vs this headache that you would choose this path. But nonetheless, my aching head thanks you <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.</p>

<p>By now you have probably worked out that <tt>IBytesRefIterator</tt> should inherit <tt>IEnumerator<T></tt> and that implementing that interface isn't much more complicated than dividing the 1 <tt>BytesRef Next();</tt> method into <tt>bool MoveNext();</tt> and <tt>BytesRef Current { get;</tt>}. The other fun stuff such as <tt>SeekCeil()</tt> and <tt>SeekExact()</tt> most likely won't be affected. And yes, this means the <tt>Dispose()</tt> method should be a no-op in our case, but who knows, maybe it won't be in a subclass somewhere - just make sure to put in the dispose pattern so we are covered.</p>

<p>The only real trick is ensuring the other methods on <tt>TermsEnum</tt> and its subclasses are exposed when implementing the <tt>IEnumerable<BytesRef></tt> on <tt>Terms</tt> (and anybody else using <tt>TermsEnum</tt>). As I mentioned before, I think overloading <tt>GetEnumerator()</tt> with a method that returns the <tt>TermsEnum</tt> type is the solution there. But, that probably doesn't take into account any subclasses of <tt>Terms</tt> that have a different type of iterator returned from their <tt>GetIterator()</tt> method than the abstract class <tt>TermsEnum</tt>.</p>

<blockquote><p>Lastly, there's the observation that the while vs foreach examples you gave are not equivalent as the while version does not call Dispose(). </p></blockquote>

<p>That is just because <tt>TermsEnum</tt> doesn't implement <tt>IDisposable</tt> yet. Of course, that all changes when implementing <tt>IEnumerator<T></tt>...</p>

<p>And there are several places throughout the codebase where enumerators are utilized without a using block or calling <tt>Dispose()</tt>. The tests aren't such a big concern, but we really shouldn't be doing this in the production code. I am sure there are some places where you can put your own implementation with its own unmanaged resources and watch the resource leaks pile up because we aren't doing our job and calling <tt>Dispose()</tt>. The real trick is locating them. I really wish Microsoft would put a feature in Visual Studio that searches for code that should exist but doesn't <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.</p>

<p>Then again, I haven't been using Code Analysis. Maybe that (or Resharper?) can be used to quickly detect all of the <tt>Dispose()</tt> violations. If you can find them, maybe that is a PR that should go ahead of the refactoring.</p>

<blockquote><p>It's odd for a type to be disposing it's "parent" (though I know some StreamWriter/Reader classes do, it's still odd/annoying).</p></blockquote>

<p>Agreed - when using <tt>Dispose()</tt> in conjunction with the decorator pattern things get odd. But how does that apply to <tt>TermsEnum</tt>? Disposing a superclass is not quite the same thing (and is rather expected using the dispose pattern).</p>

<blockquote><p>All very Tricky...</p>

<p>I think there's some combination of</p>

<ol>
	<li>isolating the "item" type</li>
	<li>using the collection object behaviors from inside the loop</li>
	<li>implementing the enumerator such that it is ok about the cursor being changed from outside</li>
</ol>


<p>It's going to be very hard to get the semantics of each type in a sane shape such that it's hard for a dev to get tripped up.</p></blockquote>

<p>In Java the "item" type is usually the one that is returned from the <tt>Next()</tt> method. But the real issue is all of the "extras" that come along with these iterators and how to expose them without casting the result of <tt>GetEnumerator()</tt>. Keep in mind the "extras" (ala LINQ) are usually part of the <tt>IEnumerable<T></tt> not part of the <tt>IEnumerator<T></tt>. Maybe trying to expose the "extras" is the wrong approach altogether. Maybe those "extras" are crying out to be refactored into services of their own that are separate from the enumerator.</p>

<p>If you go that route, note that because we are a port we had to redefine what "maintainable" means. So, when refactoring a class into many, the typical approach I follow is to drop all of that code into the same code file that is being refactored (in the same order as much as possible). This will make it easy to figure out where that code is when upgrading to the next Lucene version. The standard approach of putting each type into a file on its own would be chaos for a port version upgrade.</p>

<p>But you seem to have a handle on things. If there is a solution to be found here, I am sure you will be able to find it. If you need a sounding board to run your ideas by, I am here.</p>

<blockquote><p>Which is why this have been open for 5+ years </p></blockquote>

<p>Yea, I bet the decision to wait until the "next release" was a decision that came with some regret. Unfortunately, Lucene.Net struggled to get over the hump of getting a 4.x release because the project grew by a factor of 10. We are almost there now, and the path to catch up to Lucene 6.x is now clear considering how much smaller the change set is between here and there than the monster change set between 3.x and 4.x.</p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<p>Yeah, well, I thought, that's looks like it might be isolated, small. Ha!<br/>
I'll poke at it some more, see what turns up. But I may just punt and look for something else <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I'm not sure I'm on the same wave length just yet...</p>

<p>"how does that apply to TermsEnum? Disposing a superclass is not quite the same thing"<br/>
But the enumerator is not a super class. It's a "child" of the TermsEnum (or whatever) created for a purpose. Disposing the child shouldn't dispose the "parent". I'm fairly sure that IEnumerator has Dispose so that things like listing files (managed resource) can be cleaned up. But still, it's about cleaning up resources used by the enumerator <em>not</em> by the parent collection type.<br/>
If you want to Dispose the collection type then that's a separate operation. There may be other behaviors available, appropriate on the collection type <em>after</em> you've looped some aspect of it.<br/>
Right ??</p>

<p>Also a bit confused around the "extras" topic. I don't see the problem. We'd get all the linq bits for free. You're right, they are generally defined on <tt>IEnumerable<T></tt> But that's fine. In your example of <tt>TermsEnum</tt> being <tt>IEnumerable<BytesRef></tt> then regardless of how the enumerator is factored you can still do <tt>foreach(var item in termsEnum.Where(b=>b.Length>100))</tt></p>

<p>I was thinking by "extras" you meant "other things on the collection type". By my theory those are either: part of the item and exposed there; or an op you call in the collection object from inside the loop.</p>

<p>What am I missing?</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<p>Ahh...I think I now know where we are getting our wires crossed.</p>

<p>You have become a victim of the Java developers' incessant need to abbreviate everything (which annoys the crap out of me, but unfortunately fixing it is a losing battle). </p>

<p>Essentially, <tt>TermsEnum</tt> == <tt>IEnumerator<BytesRef></tt>, <tt>TermsEnum</tt> != <tt>IEnumerable<BytesRef></tt>. "Enum" could be interpreted as either one, hence the misunderstanding (hell, come to think of it, it could even be misinterpreted as an enumeration).</p>

<p>So, rather than loading up the <tt>IEnermerable<BytesRef></tt> with all kinds of useful stuff like LINQ does, our Lucene friends have loaded up a <tt>IEnumerator<T></tt> (or not - they didn't actually implement Java's analogue to this) with stuff that most <tt>IEnumerator</tt> s don't do. Which of course is leaving us scratching our heads.</p>

<p>So, when disposing <tt>TermsEnum</tt>, we are doing the standard <tt>IEnumerator<T>.Dispose()</tt> stuff that the foreach loop does not some funky other relationship. The <tt>Terms</tt> class is essentially the <tt>IEnumerable<BytesRef>></tt> that we are enumerating over.</p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<p>Yes, agreed. I think I wasn't quite using the right words to disambiguate.</p>

<p>So... take a breath... lots of Esomething words coming....</p>

<p>Yes, the Enum types have the characteristics of a c# IEnumerator (Next() etc).<br/>
But also represent the current item (via Term etc).<br/>
But also has additional "collection" type operations (<tt>SeekExact()</tt> etc) (are these part of what you were referring to as "extras"?).<br/>
It's this Enum class I was calling the "collection" type 'cos it kind of holds the collection state.</p>

<p>What I was proposing was to make the Enum also look like an IEnumerable so that you can <tt>foreach</tt> over it.<br/>
Which means it needs a <tt>GetEnumerable()</tt> to return an IEnumerator. This new enumerator is the <tt>EnumEnumerator</tt> type from in previous messages.</p>

<p>So, I'm taking an Enum, adding an IEnumerable facade, to return a new IEnumerable type which pokes the Enum to move forward.</p>

<p>It was the <tt>Dispose()</tt> of this new IEnumerator that I was referring to. Pretty convinced that this one should always be a no-op. Exiting the loop should not Dispose the Enum (ie the "collection") and as the new IEnumerator is just poking methods on the Enum, there's nothing for it to clean up.</p>

<p>Pretty sure I've got all the right E's in the right place <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>NB: some thought would be needed to define what the "item" class (representing Current) should look like in each case.<br/>
NB: Calling a Current changing method (ie one of the Seek...() methods) from inside a foreach would result in undefined behavior.</p>

<p>Phew... Did that make sense; fit with your view of the universe?</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<blockquote><p>It's this Enum class I was calling the "collection" type 'cos it kind of holds the collection state.</p>

<p>What I was proposing was to make the Enum also look like an IEnumerable so that you can foreach over it.<br/>
Which means it needs a GetEnumerable() to return an IEnumerator. This new enumerator is the EnumEnumerator type from in previous messages.</p>

<p>So, I'm taking an Enum, adding an IEnumerable facade, to return a new IEnumerable type which pokes the Enum to move forward.</p></blockquote>

<p>That is an interesting idea.</p>

<p>I am not sure if the state is actually being "iterated over" so much as it is being calculated on the fly. But then, <tt>IEnumerable<T></tt> is just an interface - it doesn't actually define any behavior. In fact, I had to redefine what the key and value <tt>ICollection<T></tt> s do in CharArrayMap's KeyCollection and ValueCollection, so maybe this is also possible.</p>

<p>> NB: some thought would be needed to define what the "item" class (representing Current) should look like in each case.<br/>
> NB: Calling a Current changing method (ie one of the Seek...() methods) from inside a foreach would result in undefined behavior.</p>

<p>So then we end up with some behavior that doesn't fit the mold.</p>

<p>I thought for a moment that it could be possible these SeekCeil and SeekExact methods are only defined here because there are no such thing as extension methods in Java, but since they are overridden in subclasses, that is probably not a valid assumption.</p>

<p>Speaking of which, I suggest you have a look at how they are used in the TermsEnum overloads in the codecs (for example <span class="error">&#91;Lucene40TermVectorsReader&#93;</span>(<a href="https://github.com/apache/lucenenet/blob/468199e3fa95c7e1f77b14e7f00aeaafd7c2f8b9/src/Lucene.Net/Codecs/Lucene40/Lucene40TermVectorsReader.cs#L446" class="external-link" rel="nofollow">https://github.com/apache/lucenenet/blob/468199e3fa95c7e1f77b14e7f00aeaafd7c2f8b9/src/Lucene.Net/Codecs/Lucene40/Lucene40TermVectorsReader.cs#L446</a>)) or <span class="error">&#91;Lucene45DocValuesProducer.TermsEnumAnonymousInnerClassHelper&#93;</span>(<a href="https://github.com/apache/lucenenet/blob/468199e3fa95c7e1f77b14e7f00aeaafd7c2f8b9/src/Lucene.Net/Codecs/Lucene45/Lucene45DocValuesProducer.cs#L1090" class="external-link" rel="nofollow">https://github.com/apache/lucenenet/blob/468199e3fa95c7e1f77b14e7f00aeaafd7c2f8b9/src/Lucene.Net/Codecs/Lucene45/Lucene45DocValuesProducer.cs#L1090</a>).</p>

<p>As you can see, the complexity of these enums becomes quite high (lots of internal state), and it seems to me it would be simpler to port future changes from Lucene if the general structure of the "extras" (that is, the properties and Seek* methods) were left alone (or at least left intact and used by some other type of facade). The DocsAndPositionsEnum s are even more complex than this.</p>

<p>The only thing I was hoping to get out of this was the added ability to use a foreach loop on Terms (and possibly the DocIdSetIterator). Sure we could refactor the TermsEnum data structure to be more .NET-like, but then where would we be when we need to connect the dots on the next Lucene upgrade? I am not necessarily against it, but finding bugs in the codec code is no picnic, so we would need some clear way to ensure future codecs behavior can be mapped onto the new structure without causing too much pain. Keep in mind, the codecs change in every major (and nearly every minor) version.</p>

<blockquote><p>Phew... Did that make sense; fit with your view of the universe?</p></blockquote>

<p>Wow, do I sound so closed minded? I think I need to work on my delivery <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>. </p>
<i>by nightowl888</i>

**Comment by bongohrtech** [comment]
<blockquote><p>Wow, do I sound so closed minded?</p></blockquote>
<p>Ha. No not at all <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> more worried I just not seeing the whole picture.</p>

<blockquote><p>not sure if the state is actually being "iterated over" ... </p></blockquote>
<p>Well, in so far as <tt>Next</tt> moves a set of values one step forward, it can certainly be thought of that way. I think of these like a cursor over a sql table. You can foreach over the rows. How the data is actually retrieved is abstracted. It could be one at a time, batches, all of them, cached... At the foreach level you just call Next to inspect the values of the next item.</p>

<blockquote><p><tt>IEnumerable<T></tt> is just an interface - it doesn't actually define any behavior.</p></blockquote>
<p>Right, this is one of those language syntax enabling things. You actually don't need to specify the interface. As long as the class implements <tt>GetEnumerator()</tt> then foreach will work. </p>

<blockquote><p>So then we end up with some behavior that doesn't fit the mold.</p></blockquote>
<p>True. But that's completely normal.<br/>
Imagine an enumerator over a Stream that pulls a line of text at a time. What would happen if you called <tt>myStream.Seek(...)</tt> inside the loop? Or a foreach over a Dictionary and call <tt>myDictionary.Add(...)</tt>? The behavior is to either throw (because the Dictionary.Enumerator protects against this) or some undefined behavior depending on which Stream/Reader or underlying storage is being used. Having overridden implementations with different behaviors is also to be expected. Many hours of fun to be had following Stream/TextReader etc. There's no guarantee that descendant classes will have similar behavior at he edges.</p>

<p>My theory was to enable foreach that would produce items representing the values of the Enum that change when its Next is called. This would be entirely an overlay/facade. All the existing members would remain presenting a dual api.<br/>
This way copy/pasting code from Java is still possible (modulo c# style properties). But the while style can be refactored to foreach. Or green field dev can be done in whichever style the dev likes.</p>

<p>All linq style stuff will work. I don't think the "extra" (Seek etc) are a problem. Most of them aren't composable as they return a single item or perfrom some op over the "collection". I mean for example you probably wouldn't call <tt>SeekExact</tt> from inside a foreach, in the same way you would use a dictionary indexer if you were foreach'ing over it. Or call <tt>Sort</tt> inside a foreach over a List.<br/>
Those that are composable (ie the various <tt>Docs</tt> methods) will work (assuming the <tt>DocsEnum</tt> is converted. So you could image <tt>terms.Docs(...).Where(d=>...).Select(...)</tt> working. But I would expect them to be at the beginning of the chain.<br/>
I guess you could image a few extra linq style extensions for special cases. But once you're into enumerable land it does cause some restrictions. Unless you want to get into rewriting expressions trees (runs away screaming).</p>

<p>You are quite right, trying to fit this in such that it's consistent across the class hierarchies is going to be a challenge for sure <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I'm away this weekend but I'll try to put together a PR with a few examples/tests new week to see if it'll work out.</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<p>GitHub user AndyPook opened a pull request:</p>

<p>    <a href="https://github.com/apache/lucenenet/pull/212" class="external-link" rel="nofollow">https://github.com/apache/lucenenet/pull/212</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/GH-469" title="Convert Java Iterator classes to implement IEnumerable<T>" class="issue-link" data-issue-key="GH-469">GH-469</a> - enumerables</p>

<p>    See <a href="https://issues.apache.org/jira/projects/LUCENENET/issues/GH-469" class="external-link" rel="nofollow">https://issues.apache.org/jira/projects/LUCENENET/issues/GH-469</a> for background.</p>

<p>    This is an early stage proposal for implementing IEnumerable<T> on various "Enum" types. To enable `foreach` and linq style operators.</p>

<p>    This prototype provides `EnumEnumerator<T>` as a helper. It can be used in various ways to make the implementation of `IEnumerable<T>` more straightforward.</p>

<p>    I have modified `IBytesRefIterator` and `TermsEnum` to add `IEnumerable<BytesRef>` and therefore to all of it's descendants.</p>

<p>    The test (such as it is) demonstrates retrieving a `TermsEnum` from a reader. Then using `foreach` and a simple `.Count()` linq operator.</p>

<p>    If this approach is acceptable it ought to be adaptable to other Enum types</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/AndyPook/lucenenet" class="external-link" rel="nofollow">https://github.com/AndyPook/lucenenet</a> <a href="https://issues.apache.org/jira/browse/GH-469" title="Convert Java Iterator classes to implement IEnumerable<T>" class="issue-link" data-issue-key="GH-469">GH-469</a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/lucenenet/pull/212.patch" class="external-link" rel="nofollow">https://github.com/apache/lucenenet/pull/212.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #212</p>

<hr />
<p>commit 008df498bcba7e9940ca97efc83d656ddb03b2c3<br/>
Author: Andy Pook <andy.pook@gmail.com><br/>
Date:   2017-08-14T21:32:45Z</p>

<p>    simple editorconfig</p>

<p>    ensures spaces for indent</p>

<p>commit 59646be0cf8fbc7e8ad64f395f451a11aca57fc6<br/>
Author: Andy Pook <andy.pook@gmail.com><br/>
Date:   2017-08-14T21:34:48Z</p>

<p>    add EnumEnumerator helper</p>

<p>commit 526ddaf8c837f82cd141324e922372155914701b<br/>
Author: Andy Pook <andy.pook@gmail.com><br/>
Date:   2017-08-14T21:35:32Z</p>

<p>    add IEnumerable to TermsEnum/IBytesRefArray</p>

<p>commit 456df8e8347ef67055b712722b50cc3b1bbf26cc<br/>
Author: Andy Pook <andy.pook@gmail.com><br/>
Date:   2017-08-14T21:36:54Z</p>

<p>    add header, fix usings</p>

<p>commit 1a1c25da7cbcd112d12d6c418c669786cdf4088c<br/>
Author: Andy Pook <andy.pook@gmail.com><br/>
Date:   2017-08-14T22:25:15Z</p>

<p>    add example test</p>

<hr />
<i>by githubbot</i>

**Comment by bongohrtech** [comment]
<p>added PR <a href="https://github.com/apache/lucenenet/pull/212" class="external-link" rel="nofollow">https://github.com/apache/lucenenet/pull/212</a> which I am sure will need a lot of work. The intent at this point is to just demonstrate the approach</p>
<i>by andypook</i>

**Comment by bongohrtech** [comment]
<p>Removing fix version of 4.8. This is a general lucene.net task and we should not delay 4.8 release.</p>
<i>by laimis</i>
